kion/internal/kionclient/helper.go

```diff
@@ -487,3 +488,75 @@ func PrintHCLConfig(config string) {
 	fmt.Println("Generated HCL configuration:")
 	fmt.Println(config)
 }
+
+// ConvertInterfaceSliceToIntSlice converts a slice of interfaces to a slice of integers.
+func ConvertInterfaceSliceToIntSlice(input []interface{}) []int {
+	// Create a slice of integers with the same length as the input slice.
+	output := make([]int, len(input))
+	// Iterate over the input slice, casting each element to an integer.
+	for i, v := range input {
+		output[i] = v.(int)
+	}
+	return output
+}
+
+// GetPreviousUserAndGroupIds retrieves the previous state of user and user group IDs
+// from the Terraform resource data.
+func GetPreviousUserAndGroupIds(d *schema.ResourceData) ([]int, []int) {
+	var prevUserIds, prevUserGroupIds []int
+
+	// Check if the "user_ids" field has changed
+	if d.HasChange("user_ids") {
+		// Get the previous value of the "user_ids" field
+		oldValue, _ := d.GetChange("user_ids")
+		// Convert the previous value to a slice of integers
+		prevUserIds = ConvertInterfaceSliceToIntSlice(oldValue.([]interface{}))
+	}
+
+	// Check if the "user_group_ids" field has changed
+	if d.HasChange("user_group_ids") {
+		// Get the previous value of the "user_group_ids" field
+		oldValue, _ := d.GetChange("user_group_ids")
+		// Convert the previous value to a slice of integers
+		prevUserGroupIds = ConvertInterfaceSliceToIntSlice(oldValue.([]interface{}))
+	}
+
+	return prevUserIds, prevUserGroupIds
+}
+
+// FindIdDifferences finds the differences between two slices of integers,
+// returning the elements that are present in slice1 but not in slice2.
+func FindIdDifferences(slice1, slice2 []int) []int {
+	// Create a set from the second slice for efficient lookups
+	set := make(map[int]bool)
+	for _, v := range slice2 {
+		set[v] = true
+	}
+
+	var diff []int
+	// Iterate over the first slice and find elements not present in the second slice
+	for _, v := range slice1 {
+		if !set[v] {
+			diff = append(diff, v)
+		}
+	}
+	return diff
+}
```

Comment:
This would be a good candidate for using generics so that we don't end up having to write an `int` version of this, ` string` version of this, and so on.

Reply:
FWIW, this helper already existed; it was just centralized in `[resource_project_enforcement.go](https://github.com/kionsoftware/terraform-provider-kion/blob/dcfe3354ddceb3ebe1cb1f241ac2df4f696de10c/kion/resource_project_enforcement.go#L438)`, and I moved it out of there so that I could use it here as well.  I'll look into generics!